### PR TITLE
feat: add syscalls using Barreto-Naehrig (BN) curve construction

### PR DESCRIPTION
**Pull request by Flouse**
Introduce a [pairing cryptography](https://en.wikipedia.org/wiki/Pairing-based_cryptography) library to provide [the functions](https://github.com/nervosnetwork/godwoken/blob/1b9de5231b0efdac1e56fd82a2d734dd8bbccf60/crates/generator/src/syscalls/bn.rs) of Barreto-Naehrig (BN) curve such as:

```rust
bn::add
bn::mul
bn::pairing
```

### CODE REVIEW

**Review by zeroqn — DISMISSED** [review]

File: crates/generator/src/syscalls/mod.rs [hunk 1]
```diff
@@ -524,6 +529,76 @@ impl<'a, 'b, S: State, C: ChainView, Mac: SupportMachine> Syscalls<Mac>
...
+                let input_size = machine.registers()[A4].to_u64();
+                let input = load_bytes(machine, input_addr, input_size as usize)?;
+
+                if let Some(cycles_pool) = self.cycles_pool {
+                    // k is the number of pairings being computed
+                    let k: u64 = input_size / 192;
+                    if k > 0 {
+                        // Subtract additional cycles per pairing
+                        // https://eips.ethereum.org/EIPS/eip-1108
+                        let additional_cycles = k * 34000 * 3;
```
Comment (jjyr) [inline crates/generator/src/syscalls/mod.rs]: Can we add this modifier factor to `SyscallCyclesConfig`?

File: crates/config/src/config.rs [hunk 2]
```diff
@@ -560,10 +560,13 @@ pub struct SyscallCyclesConfig {
...
     pub sys_recover_account_cycles: u64,
     pub sys_log_cycles: u64,
+    pub sys_bn_add_cycles: u64,
+    pub sys_bn_mul_cycles: u64,
+    pub sys_bn_pairing_cycles: u64,
 }
 
 impl SyscallCyclesConfig {
-    pub fn all_zero() -> Self {
+    pub fn default() -> Self {
```
Comment (jjyr) [inline crates/config/src/config.rs:579]: Please modify the default config according to mainnet settings.

Comment (Flouse) [reply crates/config/src/config.rs:579]: https://github.com/godwokenrises/godwoken/pull/808/commits/ead4e4faf3b36f0f5e08df9fa51600f774350e4b

File: crates/generator/src/syscalls/mod.rs (hunk 1 above)
Comment (Flouse) [reply crates/generator/src/syscalls/mod.rs]: https://github.com/godwokenrises/godwoken/pull/808/commits/61b58b1d232570f054345c55c7fcf821338607c8

**Comment by jjyr** [comment]
Please fix the CI

**Review by jjyr — APPROVED** [review]

**Review by magicalne — APPROVED** [review]

**Pull request merged**